qiskit/pulse/schedule.py

```diff
@@ -650,23 +665,33 @@ def assign_parameters(self,
         Returns:
             Schedule with updated parameters (a new one if not inplace, otherwise self).
         """
-        for _, inst in self.instructions:
-            inst.assign_parameters(value_dict)
+        for parameter in self.parameters:
+            if parameter not in value_dict:
+                continue
+
+            value = value_dict[parameter]
+            for inst in self._parameter_table[parameter]:
+                inst.assign_parameters({parameter: value})
 
+            entry = self._parameter_table.pop(parameter)
+            if isinstance(value, ParameterExpression):
+                for new_parameter in value.parameters:
+                    if new_parameter in self._parameter_table:
+                        new_entry = set(entry + self._parameter_table[new_parameter])
+                        self._parameter_table[new_parameter] = list(new_entry)
+                    else:
+                        self._parameter_table[new_parameter] = entry
+
+        # Update timeslots according to new channel keys
         for chan in copy.copy(self._timeslots):
             if isinstance(chan.index, ParameterExpression):
                 chan_timeslots = self._timeslots.pop(chan)
 
                 # Find the channel's new assignment
                 new_channel = chan
-                for param in chan.index.parameters:
-                    if param in value_dict:
-                        new_index = new_channel.index.assign(param, value_dict[param])
-                        if not new_index.parameters:
-                            new_index = float(new_index)
-                            if float(new_index).is_integer():
-                                new_index = int(new_index)
-                        new_channel = type(chan)(new_index)
+                for param, value in value_dict.items():
+                    if param in new_channel.parameters:
+                        new_channel = new_channel.assign(param, value)
```

Comment:
What happens if the old channel was used in an instruction? Doesn't this produce a new channel that isn't used anywhere except in timeslots?

Reply:
I don't know what you are getting at. `DriveChannel(0)` is equal to `DriveChannel(0)`; it doesn't matter if they are the same instance